api/app/Policies/UserPolicy.php

```diff
@@ -80,11 +82,72 @@ public function updateSub(User $user)
     /**
      * Determine whether the user can update roles.
      *
+     * @param  UpdateUserRolesInput  $args
      * @return \Illuminate\Auth\Access\Response|bool
      */
-    public function updateRoles(User $user)
+    public function updateRoles(User $user, $args)
     {
-        return $user->isAbleTo('assign-any-role');
+        $targetUserId = isset($args['id']) ? $args['id'] : null;
+        $attachRoles = isset($args['roleAssignmentsInput']) && isset($args['roleAssignmentsInput']['attach']) ?
+            $args['roleAssignmentsInput']['attach'] : null;
+        $detachRoles = isset($args['roleAssignmentsInput']) && isset($args['roleAssignmentsInput']['detach']) ?
+            $args['roleAssignmentsInput']['detach'] : null;
+
+        if (is_null($targetUserId)) {
+            return false;
+        }
+
+        // adding or removing team based roles
+        if (isset($attachRoles['team']) || isset($detachRoles['team'])) {
```

Comment:
I take back my previous comment about decoupling this policy from the schema. However, if we follow my suggested change to the schema, then it probably makes more sense to simply loop through all the attach roles then all the detach roles, instead of dealing with team roles first then non-team roles.

Reply:
Adjusting the policy to fit the new mutation input type simplifies the policy file a fair bit
https://github.com/GCTC-NTGC/gc-digital-talent/pull/10791/commits/498008038a3e1857fae7e57711346eb42c61ba6f